source/Week_11_correction/Graph avec des classes - solution.ipynb

```diff
@@ -0,0 +1,180 @@
+    "+ `del_edge` pour supprimer un `edge`\n",
+    "+ `__repr__` pour retourner l'objet sous forme de dictionnaire. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "class Edge:\n",
+    "    # votre code ici\n",
+    "    def __init__(self, from_vertex, to_vertex, weight):\n",
+    "        self.from_vertex = from_vertex\n",
+    "        self.to_vertex = to_vertex\n",
+    "        self.weight = weight\n",
+    "\n",
+    "    def __repr__(self):\n",
+    "        return str({\"from_vertex\": self.from_vertex, \"to_vertex\": self.to_vertex, \"weight\": self.weight})\n",
+    "\n",
+    "    \n",
+    "class Graph:\n",
+    "    # votre code ici\n",
+    "    def __init__(self):\n",
+    "        self.edges = set()\n",
+    "        self.vertices = set()\n",
+    "\n",
+    "    def add_vertex(self, name):\n",
+    "        self.vertices.add(name)\n",
+    "\n",
+    "    def edge_exist(self, from_vertex, to_vertex):\n",
+    "        for edge in self.edges:\n",
+    "            if edge.from_vertex is from_vertex and edge.to_vertex is to_vertex:\n",
+    "                return edge.weight\n",
+    "        return False\n",
+    "\n",
+    "    def __generate_edge(self, from_vertex, to_vertex, weight):\n",
+    "        if from_vertex in self.vertices and to_vertex in self.vertices:\n",
+    "            new_edge = Edge(from_vertex, to_vertex, weight)\n",
+    "            self.edges.add(new_edge)\n",
+    "            return\n",
+    "        else:\n",
+    "            if to_vertex not in self.vertices:\n",
+    "                self.add_vertex(to_vertex)\n",
+    "            if from_vertex not in self.vertices:\n",
+    "                self.add_vertex(from_vertex)\n",
+    "            new_edge = Edge(from_vertex, to_vertex, weight)\n",
+    "            self.edges.add(new_edge)\n",
+    "\n",
+    "    def update_weight(self, from_vertex, to_vertex, new_weight):\n",
+    "        for edges in self.edges:\n",
+    "            if edges.from_vertex is from_vertex and edges.to_vertex is to_vertex:\n",
+    "                edges.weight = new_weight\n",
+    "                print(\"weight has been updated\")\n",
+    "                return\n",
+    "        self.new_edge(from_vertex, to_vertex, new_weight)\n",
+    "        print(\n",
+    "            \"The vertex between {} and {} does not exist and will be created\".format(\n",
+    "                from_vertex, to_vertex))\n",
+    "\n",
+    "    def new_edge(self, from_vertex, to_vertex, weight):\n",
+    "        test_existence = self.edge_exist(from_vertex, to_vertex)\n",
+    "        if test_existence is not False:\n",
```

Comment:
Could rewrite like:

        if test_existence:
            if test_existence is weight:
                print("Edge between {} and {} with the same weight already exist".format(from_vertex, to_vertex))
            else:
                print("Edge between {} and {} with the same weight already exist but with a"
                      "different weight and will be overwritten".format(from_vertex, to_vertex))
                self.update_weight(from_vertex, to_vertex, weight)
        else:
            self.__generate_edge(from_vertex, to_vertex, weight)

Reply:
much cleaner I will update thanks !